File: go/extensions/bazaar/startup_validation.go
package bazaar

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"github.com/x402-foundation/x402/go/v2/extensions/types"
	x402http "github.com/x402-foundation/x402/go/v2/http"
)

var httpVerbRE = regexp.MustCompile(`(?i)^(GET|POST|PUT|PATCH|DELETE|HEAD)\b`)

// withSyntheticMethod injects a synthetic method for startup schema validation only.
//
// Pre-enrichment HTTP extensions intentionally omit method; it is added at request
// time by BazaarResourceServerExtension. Without a synthetic value, jsonschema reports
// a false positive for the required method field.
//
// Priority: (1) route pattern verb (e.g. "GET /api"), (2) body vs query inference.
// Returns the same extension unchanged if method is already present.
func withSyntheticMethod(ext types.DiscoveryExtension, pattern string) types.DiscoveryExtension {
	switch input := ext.Info.Input.(type) {
	case types.QueryInput:
		if input.Method != "" {
			return ext
		}
		input.Method = types.QueryParamMethods(inferSyntheticMethod(pattern, input))
		ext.Info.Input = input
	case types.BodyInput:
		if input.Method != "" {
			return ext
		}
		input.Method = types.BodyMethods(inferSyntheticMethod(pattern, input))
		ext.Info.Input = input
	}
	return ext
}

func inferSyntheticMethod(pattern string, input interface{}) string {
	if matches := httpVerbRE.FindStringSubmatch(pattern); len(matches) > 1 {
		return strings.ToUpper(matches[1])
	}

	switch in := input.(type) {
	case types.BodyInput:
		if in.Body != nil || in.BodyType != "" {
			return "POST"
		}
	case types.QueryInput:
		// Query-only extensions default to GET below.
	}

	return "GET"
}

// ValidateBazaarRouteExtensions validates bazaar extensions on all routes using
// JSON-schema validation. Emits warnings for invalid extensions but does not block startup.
func ValidateBazaarRouteExtensions(routes x402http.RoutesConfig) {
	for pattern, config := range routes {
		validateSingleBazaarExtension(pattern, config.Extensions)
	}
}

// ValidateBazaarRouteExtensionsFromServer validates bazaar extensions from pre-compiled routes.
func ValidateBazaarRouteExtensionsFromServer(server *x402http.HTTPServer) {
	for _, route := range server.GetCompiledRoutes() {
		pattern := route.Verb + " " + route.Regex.String()
		validateSingleBazaarExtension(pattern, route.Config.Extensions)
	}
}

func validateSingleBazaarExtension(pattern string, extensions map[string]interface{}) {
	extVal, ok := extensions[types.BAZAAR.Key()]
	if !ok || extVal == nil {
		return
	}

	switch v := extVal.(type) {
	case map[string]interface{}:
		if v["info"] == nil || v["schema"] == nil {
			fmt.Printf("x402 Warning: Route %q declares a bazaar extension but it is malformed "+
				"(expected an object with \"info\" and \"schema\" fields)\n", pattern)
			return
		}
	case types.DiscoveryExtension:
		if v.Schema == nil {
			fmt.Printf("x402 Warning: Route %q declares a bazaar extension but it is malformed "+
				"(expected an object with \"info\" and \"schema\" fields)\n", pattern)
			return
		}
	default:
		fmt.Printf("x402 Warning: Route %q declares a bazaar extension but it is malformed "+
			"(expected an object with \"info\" and \"schema\" fields)\n", pattern)
		return
	}

	extJSON, err := json.Marshal(extVal)
	if err != nil {
		return
	}
	var ext types.DiscoveryExtension
	if err := json.Unmarshal(extJSON, &ext); err != nil {
		return
	}
	specResult := ValidateDiscoveryExtensionSpec(ext)
	if !specResult.Valid {
		fmt.Printf("x402 Warning: Route %q has invalid bazaar extension: %s\n",
			pattern, strings.Join(specResult.Errors, ", "))
		return
	}
	result := ValidateDiscoveryExtension(withSyntheticMethod(ext, pattern))
	if !result.Valid {
		fmt.Printf("x402 Warning: Route %q has invalid bazaar extension: %s\n",
			pattern, strings.Join(result.Errors, ", "))
	}
}

File: typescript/packages/extensions/src/bazaar/startupValidation.ts
/**
 * Shared startup-time validation utilities for bazaar extensions in route configs.
 *
 * Used by middleware packages (Express, Hono, Next) to validate bazaar extensions
 * at server startup without duplicating the iteration and warning logic.
 */

import type { RoutesConfig } from "@x402/core/server";
import type { DiscoveryExtension } from "./types";
import { validateDiscoveryExtension, validateDiscoveryExtensionSpec } from "./facilitator";

export { checkIfBazaarNeeded } from "@x402/core/server";

const HTTP_VERB_RE = /^(GET|POST|PUT|PATCH|DELETE|HEAD)\b/i;

/**
 * Inject a synthetic method into a pre-enrichment extension so the schema's
 * required:["method"] check doesn't produce a false-positive warning at startup.
 * Priority: (1) route pattern verb (e.g. "GET /api"), (2) body vs query inference.
 * Returns the same object unchanged if method is already present.
 *
 * @param ext - The raw bazaar extension object
 * @param pattern - The route pattern key (e.g. "GET /api" or "*")
 * @returns The extension with a synthetic method injected into info.input if needed
 */
function withSyntheticMethod(
  ext: Record<string, unknown>,
  pattern: string,
): Record<string, unknown> {
  const info = ext.info as Record<string, unknown> | undefined;
  const input = info?.input as Record<string, unknown> | undefined;
  if (!input || (typeof input.method === "string" && input.method)) {
    return ext;
  }
  const verbMatch = pattern.match(HTTP_VERB_RE);
  const method = verbMatch
    ? verbMatch[1].toUpperCase()
    : input.body !== undefined || input.bodyType !== undefined
      ? "POST"
      : "GET";
  return { ...ext, info: { ...info, input: { ...input, method } } };
}

/**
 * Validate bazaar extensions on all routes using JSON-schema validation.
 * Emits console warnings for invalid extensions but does not throw.
 *
 * @param routes - Route configuration to scan for bazaar extensions
 */
export function validateBazaarRouteExtensions(routes: RoutesConfig): void {
  const entries: [string, { extensions?: Record<string, unknown> }][] =
    "accepts" in routes ? [["*", routes]] : Object.entries(routes);

  for (const [pattern, config] of entries) {
    const bazaarExt = config.extensions?.["bazaar"];
    if (!bazaarExt) continue;
    if (
      typeof bazaarExt === "object" &&
      bazaarExt !== null &&
      "info" in (bazaarExt as Record<string, unknown>) &&
      "schema" in (bazaarExt as Record<string, unknown>)
    ) {
      const specResult = validateDiscoveryExtensionSpec(bazaarExt as Record<string, unknown>);
      if (!specResult.valid) {
        console.warn(
          `x402: Route "${pattern}" has an invalid bazaar extension: ${specResult.errors?.join(", ")}`,
        );
        continue;
      }
      const extForSchema = withSyntheticMethod(bazaarExt as Record<string, unknown>, pattern);
      const schemaResult = validateDiscoveryExtension(
        extForSchema as unknown as DiscoveryExtension,
      );
      if (!schemaResult.valid) {
        console.warn(
          `x402: Route "${pattern}" has an invalid bazaar extension: ${schemaResult.errors?.join(", ")}`,
        );
      }
    } else {
      console.warn(
        `x402: Route "${pattern}" declares a bazaar extension but it is malformed ` +
          `(expected an object with "info" and "schema" fields)`,
      );
    }
  }
}
